SEED-platform/seed
# Add UBID relation to state

#### Any background context you want to provide?
The PropertyState (and TaxLotState) both have a `ubid` field that aligns with a canonical column. Per some new guidelines States can have multiple ubids. To accommodate this a new model `UbidModel` has been created with a many-to-one relationship with PropertyState and TaxlotState. Note that the name of the new model could not be `ubid` to prevent a naming conflict with the existing field.

A UbidModel will have the following fields:
- ubid: string 
- preferred: boolean 
- property: ForeignKey
- taxlot: ForeignKey

PNNL has developed a number of sql functions that can be migrated into a postgres db. [Reference](https://github.com/pnnl/buildingid-plpgsql/tree/master). Some of these functions allow a user to calculate the overlap of 2 unique UBID's. This is refered to as the Jaccard Index and the functions returns a value from 0 to 1 representing the match quality.
![image (3)](https://github.com/SEED-platform/seed/assets/58446472/232bd424-9fef-4fec-a5af-827d768cbe57)


Now that UBIDs can be compared via the jaccard index, the matching and merging processes must be updated to use a threshold value instead of an exact match. 

#### What's this PR do?

- Creates a relationship between -State and UbidModel. -States will still have a ubid column representing the 'prefered' ubid model
- Adds backend CRUD endpoints 
- Adds SQL migration to the DB to leverage PNNL functions. The PNNL Function we are most interested in is `UBID_CodeArea_Jaccard`. This function has a number of dependent functions required to run. These dependant functions are: `UBID_Decode`, `UBID_Parse`, `UBID_CodeArea`, `UBID_ValidateLatitude`, `UBID_ValidateLongitude`, `pluscode_isfull`, `pluscode_isvalid`, `pluscode_isshort`, `pluscode_decode`, `pluscode_codearea`
- Generates a new Organization setting for a UBID threshold. The threshold is an org setting that must be between 0 and 1.0. If the jaccard index between an incoming ubid and an existing ubid is above the UBID threshold it will qualify as a match (if the other matching criteria are met)
- Uses signals to generate UbidModels when a new property or taxlot is uploaded with a ubid
- Uses signals to update -State.ubid if a saved UbidModel's "preferred" field is set to true
- Adds frontend CRUD. A new ubid viewer and editor are available from the inventory list actions, inventory detail actions, or under a new UBID inventory detail nav button. The structure is loosley based on the derived column admin and editors. New or updated ubids are validated before they may be saved.
- Adds a ubid jaccard comparison modal to the inventory list actions. It can be triggered when 0-2 properties are selected. If fewer than 2 properties are selected a user may use custom ubids for comparison.
- Renames the ubid_modal to ubid_decode_modal, which allows users to decode a property's ubid
- Allows properties to be merged together while preserving the ubid history.
- Adds a map view to the inventory detail and inventory detail ubid pages highlighting the Properties ubid bounding box. If there is no preferred ubid the map is disabled. If the preferred ubid is updated the page and map reload. 

#### How should this be manually tested?
Run tests, test for functionality.

Possible testing procedure:
- upload a few properties with ubids 
- set org setting ubid threshold to something small like 0.1 to allow matching
- upload the same properties with slightly different UBIDs (ex: 85FPPRR9+3C-0-0-0-0 --->  85FPPRR9+3C-0-0-0-1)
- properties and ubids should have been merged with the preferred set to the incoming 
- view Inventory Detail UBIDs page, verify map 
- change the preferred UBID 
- map should update 
- create a UBID, delete a UBID 
- set org setting ubid threshold to something high like 0.9 to prevent matching 
- upload the same properties with different UBIDS (ex: 85FPPRR9+3C-0-0-0-1 ---> 85FPPRR9+3C-0-0-0-5)
- property list should have created new properties as the ubid threshold was not met 
- set org setting ubid threshold to something small like 0.1
- upload same properties with different UBIDs
- existing properties should have been merged into the incoming properties
- ensure ubid history is preserved on Inventory Detail UBIDs
- merge 2 properties together, ensure ubid history
- unmerge property, check ubid history

#### What are the relevant tickets?
#3957 
#4051
#4069
#4041 
#4064 

#### Screenshots (if appropriate)
Should this page include a graphic illustrating what a jaccard match looks like (as seen in PR background)?

![Screenshot 2023-06-01 at 3 13 56 PM](https://github.com/SEED-platform/seed/assets/58446472/e47b6adb-76f3-4eed-9088-d5a9393c0f94)
![Screenshot 2023-06-01 at 3 10 26 PM](https://github.com/SEED-platform/seed/assets/58446472/3864c457-aa2c-45c0-a4e1-2b23bcfb0ba7)
![Screenshot 2023-06-01 at 3 11 09 PM](https://github.com/SEED-platform/seed/assets/58446472/6558d31b-378e-4171-ac5c-6006552af657)
![Screenshot 2023-06-08 at 12 55 13 PM](https://github.com/SEED-platform/seed/assets/58446472/5fe5bd47-9565-415b-9670-27f81b1ea200)
![Screenshot 2023-06-08 at 12 55 01 PM](https://github.com/SEED-platform/seed/assets/58446472/6bd6fd82-c83f-490e-97a1-ba18b3df4b55)
